### ISSUE TITLE
Getting Google Profile Picture to the Client

### ISSUE THREAD

**Issue by oliverw** [DuendeSoftware/Support#483]
**Which version of Duende IdentityServer are you using?**
6

**Which version of .NET are you using?**
.NET 6

**Describe the question**

I have a working setup using Google (and other) as external Identiy Providers to IdentityServer 6 and a couple of mobile and desktop clients using the authorization code flow. Now I would like the client to receive the profile picture of the user for each service. Starting with Google I added this:

```
.AddGoogle(options =>
{
   ...

    options.Events.OnCreatingTicket = (context) =>
    {                      
        var picture = context.User.GetProperty("picture").GetString();

        context.Identity.AddClaim(new Claim("picture", picture));

        return Task.CompletedTask;
    };
})

```

This code does add the picture URL to the identity claims but I am not sure how I get IS to pass this claim to the client.

I think I need to add it here but I'm not sure how:

```
.AddInMemoryIdentityResources(new[]
{
    new IdentityResources.OpenId(),
    new IdentityResources.Profile(),
    new IdentityResources.Email(),
})

```

Any suggestions?

**Comment by brockallen** [comment]
Hi @oliverw. Have you had a look at our docs on making user information available to clients?

https://docs.duendesoftware.com/identityserver/v6/fundamentals/claims/

And then clients would have to request those via scopes:

https://docs.duendesoftware.com/identityserver/v6/fundamentals/resources/

**Comment by oliverw** [comment]
> Hi @oliverw. Have you had a look at our docs on making user information available to clients?
> 
> https://docs.duendesoftware.com/identityserver/v6/fundamentals/claims/
> 
> And then clients would have to request those via scopes:
> 
> https://docs.duendesoftware.com/identityserver/v6/fundamentals/resources/

Now I did. Unfortunately my `picture` claim (see snippet above) never makes it into the `ProfileDataRequestContext` when `IProfileService.GetProfileDataAsync` is called.

**Comment by brockallen** [comment]
> Unfortunately my picture claim (see snippet above) never makes it into the ProfileDataRequestContext when IProfileService.GetProfileDataAsync is called.

This would be up to your login page design how to make that available. That's commonly done via some persistence between your external login and the profile service implementation. That's nothing that IdentityServer does automatically.

**Comment by oliverw** [comment]
> 

But isn't new `IdentityResources.Profile()` configured to include `JwtClaimTypes.Picture`? Is there no way to enrich the data returned by an external identity provider in a way for IdentityServer to pick up well known claims automatically?

**Comment by josephdecock** [comment]
I recommend taking a look at this page in the docs as well: https://docs.duendesoftware.com/identityserver/v6/ui/login/external/, which goes into detail about how external providers should be handled in general and describes how to handle claims from an external provider.

**Comment by oliverw** [comment]
Figured it out. `AutoProvisionUserAsync` in `External/Login/Callback.cshtml.cs` was the missing link where both the added claim from `OnCreatingTicket` is still available and user creation happens. So all I had to add is this:

```
var picture = claims.FirstOrDefault(x => x.Type == "picture")?.Value;
if(picture != null)
    filtered.Add(new Claim(JwtClaimTypes.Picture, picture));
```

And the `JwtClaimTypes.Picture` claim would be picked up through `IdentityResources.Profile()` further down the pipeline.
